x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Child context deadline exceeded error during dialing should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.Background(), nil
+			},
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(ctx context.Context, network, addr string) (net.Conn, error) {
+					// Wrap the context in a context with an already-exceeded deadline. Expect the
+					// error returned by DialContext() to be treated as a timeout caused by reaching
+					// connectTimeoutMS.
+					ctx, cancel := context.WithDeadline(ctx, time.Now().Add(-1*time.Hour))
+					defer cancel()
+					return d.DialContext(ctx, network, addr)
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+		{
+			desc: "Parent context deadline exceeded error during handshake should not clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(ctx context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						// Sleep for at least 150ms and expect the context passed to
+						// server.Connection() to time out during the sleep. Expect the error
+						// returned by GetHandshakeInformation() to be treated as a timeout caused
+						// by an operation-scoped deadline.
+						// E.g. FindOne(context.WithTimeout(...))
+						time.Sleep(150 * time.Millisecond)
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Child context deadline exceeded error during handshake should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(ctx context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						// Wrap the context in a context with an already-exceeded deadline. Expect
+						// the error returned by GetHandshakeInformation() to be treated as a
+						// timeout caused by connectTimeoutMS.
+						ctx, cancel := context.WithDeadline(ctx, time.Now().Add(-1*time.Hour))
+						defer cancel()
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+		{
+			desc: "Dial errors unrelated to context timeouts should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 1*time.Second)
```

Comment:
```suggestion
				return context.Background(), nil
```

Reply:
See above comment about using `context.WithTimeout` vs `context.Background` here.

I'll add a comment to clarify the intent of the `context.WithTimeout`.